### **Nama    : Tiara Fitra Ramadhani Siregar**
### **Nim     : 24060121120008**
### **Dataset : Rice Dataset Cammeo and Osmancik**
### **Link    : https://www.kaggle.com/datasets/muratkokludataset/rice-dataset-commeo-and-osmancik**

## **TUGAS NOMOR 1**

In [130]:
# Mengimpor modul Pandas dengan alias atau singkatan pd
import pandas as pd

In [131]:
# Memuat dataset Iris dari URL
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

# Menetapkan sebuah variabel bernama names yang berisi daftar nama-nama kolom yang terdapat pada dataset Iris
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']

# Menetapkan sebuah variabel bernama dataset1 berisi sebuah objek pandas DataFrame yang dibuat dari membaca file CSV di URL dan nama kolom dari names
dataset1 = pd.read_csv(url, names=names)

In [132]:
# Mengimpor modul model_selection dari library scikit-learn untuk membantu proses pemilihan dan evaluasi model machine learning
from sklearn import model_selection

**I. Membuat Validasi Dataset**

In [133]:
# Mengambil nilai-nilai dari DataFrame dataset1 dan dibuat menjadi sebuah array numpy dan disimpan pada sebuah variabel bernama array
array = dataset1.values

# Assign semua baris dan empat kolom pertama pada variable X
X = array[:,0:4]

# Assign semua baris dan kolom kelima pada variable Y
Y = array[:,4]

# Menetapkan proporsi atau persentase data yang digunakan sebagai data validasi atau uji adalah 20% dan sisanya yaitu 80% sebagai data latih
validation_size = 0.20

# Menunjukkan nilai acak atau random state yang akan digunakan untuk memastikan hasil yang konsisten dan memastikan bahwa data akan dibagi dengan cara yang sama setiap kali kode dijalankan
seed = 7

# Menetapkan empat variable yang berisi hasil dari membagi data menjadi data latih dan  data validasi
# Fungsi model_selection.train_test_split memiliki parameter X dan Y yang merupakan data fitur dan target, text_size yang merupakan proporsi dan validasi, dan random_state yang merupakan nilai acak.
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size = validation_size, random_state=seed, shuffle=True)
# Kode di atas mengembalikan empat array yaitu :
# X_train dan Y_train yang merupakan data fitur dan target untuk data latih
# X_validation dan Y_validation yang merupakan data fitur dan target untuk data validasi


**II. Mencetak Ukuran atau Dimensi Data Latih dan Data Validasi**

In [134]:
# Mencetak string "X_train" dan ukuran dari array X_train yang berisi fitur untuk data latih
print("X_train :", X_train.shape)

# Mencetak string "X_validation" dan ukuran dari array X_validation yang berisi fitur untuk data validasi
print("X_validation :", X_validation.shape)

# Mencetak string "Y_train" dan ukuran dari array Y_train yang berisi data target untuk data latih
print("Y_train :", Y_train.shape)

# Mencetak string "Y_validation" dan ukuran dari array Y_validation yang berisi data target untuk data validasi
print("Y_validation :", Y_validation.shape)

X_train : (120, 4)
X_validation : (30, 4)
Y_train : (120,)
Y_validation : (30,)


Penjelasan:

- X_train : (120, 4) menunjukkan bahwa data fitur untuk data latih memiliki 120 baris dan 4 kolom

- X_validation : (30, 4) menunjukkan bahwa data fitur untuk data validasi memiliki 30 baris dan 4 kolom

- Y_train : (120,) menunjukkan bahwa data  target untuk data latih memiliki 120 baris dan 1 kolom

- Y_validation : (30,) menunjukkan bahwa data target untuk data validasi memiliki 30 baris dan 1 kolom

**III. K-Folds Cross Valiadation**

In [135]:
# Menunjukkan nilai acak atau random state yang digunakan untuk memastikan hasil yang konsisten dan memastikan bahwa data akan dibagi dengan cara yang sama setiap kali kode dijalankan
seed1 = 7

# Metrik evaluasi yang akan digunakan untuk membandingkan model yang berbeda adalah nilai akurasi, yaitu seberapa sering prediksi model sesuai dengan kelas yang sebenarnya dari data
scoring1 = 'accuracy'

**IV. Membangun Model**

In [136]:
# Mengimpor kelas DecisionTreeClassifier dari modul sklearn.tree, kelas ini merupakan estimator untuk membuat dan melatih model klasifikasi berdasarkan decision tree
from sklearn.tree import DecisionTreeClassifier

# Mengimpor kelas RandomForestClassifier dan GradientBoostingClassifier dari modul sklearn.ensemble, kedua kelas tersebut merupakan estimator untuk membuat dan melatih sebuah model klasifikasi berdasarkan random forest dan gradientboosting
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

# Mengimpor tiga fungsi dari modul sklearn.metrics yaitu accuracy_score, confusion_matrix, dan classification_report untuk menghitung dan menammpilkan beberapa metrik evaluasi untuk model machine learning
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [137]:
# membuat sebuah variabel bernama models yang berisi sebuah daftar kosong
models1 = []

# Menambahkan sebuah elemen ke dalam daftar models yaitu tuple berisi string 'Decision Tree' dan sebuah objek dari kelas DecisionTreeClassifier
models1.append(('Decision Tree', DecisionTreeClassifier()))

# Menambahkan sebuah elemen ke dalam daftar models yaitu tuple berisi string 'Random Forest' dan sebuah objek dari kelas RandomForestClassifier
models1.append(('Random Forest', RandomForestClassifier()))

# Menambahkan sebuah elemen ke dalam daftar models yaitu tuple berisi string 'Gradient Boosting' dan sebuah objek dari kelas GradientBoostingClassifier
models1.append(('Gradient Boosting', GradientBoostingClassifier()))

# Evaluate each model in turn

# Membuat variable result kosong
results1 = []

# Membuat varible names kosong
names1 = []

# Looping setiap elemen di models dan menetapkan variable name dan model sebagai nilai dari tuple yang ada di dalam elemen tsb
for name, model in models1:
  # Membagi data menjadi 10 bagian atau folds dengan menggunakan nilai acak dari variable seed dan mengacak urutan data sebelum dibagi
  kfold = model_selection.KFold(n_splits=10, random_state=seed1, shuffle=True)

  # Membuat variable cv_result berisi sebuah array numpy yang berisi nilai akurasi dari model yang diuji dengan menggunakan metode cross-validation, dan parameter sbb:
  # - model sebagai model yang akan diuji
  # - X_train dan Y_train sebagai data fitur dan target dari data latih
  # - kfold sebagai metode pembagian data
  # - scoring sebagai metrik evaluasi yang digunakan
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring1)

  # Menambahkan nilai dari variable cv_result ke dalam daftar result
  results1.append(cv_results)

  # Menambahkan nilai dari variable ke dalam daftar names
  names1.append(name)

  # Membuat sebuah variable bernama msg yang ebrisi sebauh string berformat yang menampilkan nama model, nilai rata-rata , dan standar deviasi dari nilai akurasi yang ada di dalam variable cv_reslut
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())

  # Mencetak nilai dari variable msg ke layar
  print(msg)

Decision Tree: 0.950000 (0.076376)
Random Forest: 0.966667 (0.040825)
Gradient Boosting: 0.950000 (0.055277)


**Penjelasan:**

- Decision Tree memiliki nilai akurasi 95% dan standar deviasi 7.63%
- Random Forest memiliki nilai akurasi 96.67% dan standar deviasi 4.08%
- Gradient Boosting memiliki nilai akurasi 95% dan standar deviasi 5.52%

**Kesimpulan:**
- Random Forest memiliki nilai akurasi tertinggi dan dapat memprediksi kelas yang benar dari data validasi sebanyak 96.67% dari keseluruhan model data
- Decision Tree dan Gradient Boosting memiliki nilai akurasi yang sama dan dapat memprediksi kelas yang benar dari data validasi sebanyak 95% dari keseluruhan data
- Standar deviasi untuk Decision Tree adalah yang paling tinggi yang menunjukkan bahwa nilai akurasi bervariasi cukup besar antara setiap folds atau bagian dari data. Hal ini mengindikasikan bahwa model tidak stabil atau sensitif terhadap perubahan data.

**V. Memilih Model Terbaik**

Berdasarkan hasil skor akurasi di atas, Random Forest memiliki nilai akurasi perkiraan terbesar. Berikut merupakan hasil pengujian tentang keakuratan setiap model terhadap data yang ada.

**a. Decision Tree**

Decision Tree adalah sebuah algoritma machine learning berdasarkan pada konsep pohon keputusan, yaitu sebuah struktur yang terdiri dari simpul dan cabang yang merepresentasikan pertanyaan dan jawaban yang berhubungan dengan data. Kelebihan dari algoritma ini adalah mudah untuk diinterpretasikan, dapat menangani data yang memiliki fitur kategorikal dan numerik, dan tidak membutuhkan banyak pra-pemrosesan data. Kekurangan dari algoritma ini adalah rentan terhadap overfitting, sensitif terhadap perubahan data, dan dapat menghasilkan hasil yang berbeda jika menggunakan kriteria pemisahan yang berbeda.

In [139]:
# Membuat objek dt yang merupakan sebuah estimator dari kelas DecisionTreeClassifier
dt = DecisionTreeClassifier()

# Melatih model dt dengan menggunakan data fitur dan target dari data latih
dt.fit(X_train, Y_train)

# Menggunakan model dt yang sudah dilatih untuk membuat prediksi kelas dari data fitur validasi, yaitu X_validation dan simpan hasilnya dalam variable predictions_dt
predictions_dt = dt.predict(X_validation)

# Mencetak string judul accuracy score dari decision tree
print("Accuracy Score of Decision Tree:")

# Mencetak nilai akurasi dari model dt, yaitu seberapa sering prediksi model dt sesuai dengan kelas yang sebenarnya dari data validasi
# Nilai akurasi dihitung dengan mengambil parameter Y_validation dan predictions_dt.
print(accuracy_score(Y_validation, predictions_dt))

# Mencetak string judul confusion matrix dari decision tree
print("\nConfusion Matrix of Decision Tree:")

# Mencetak confusion matrix dari model dt yang menampilkan jumlah data yang benar terklasifikasi (diagonal utama) dan jumlah data yang salah terklasifikasi (di luar diagonal utama) untuk setiap kelas yang ada
print(confusion_matrix(Y_validation, predictions_dt))

# Mencetak string judul classification report dari decision tree
print("\nClassification Report of Decision Tree:")

# Mencetak classification report dari model dt yang menampilkan nilai-nilai metrik evaluasi untuk setiap kelas, serta rata-rata dari semua kelas
print(classification_report(Y_validation, predictions_dt))

Accuracy Score of Decision Tree:
0.8666666666666667

Confusion Matrix of Decision Tree:
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]

Classification Report of Decision Tree:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



**Penjelasan hasil:**

- Nilai akurasi adalah 0.8666666666666667, yang berarti model dt dapat memprediksi kelas yang benar untuk 87% dari data validasi

- Dari confusion matrix, dapat dilihat bahwa model dt dapat mengklasifikasikan semua data Iris-setosa dengan benar, tetapi ada beberapa kesalahan klasifikasi untuk data Iris-versicolor dan Iris-virginica. Misalnya, ada dua data Iris-versicolor yang diprediksi sebagai Iris-virginica, dan ada dua data Iris-virginica yang diprediksi sebagai Iris-versicolor.

- Dari classification report, dapat dilihat bahwa model dt memiliki nilai precision, recall, dan f1-score yang tinggi untuk kelas Iris-setosa. Nilai precision menunjukkan seberapa sering model dt memprediksi kelas yang benar dari semua prediksi yang dibuat untuk kelas tersebut. Nilai recall menunjukkan seberapa sering model dt memprediksi kelas yang benar dari semua data yang sebenarnya memiliki kelas tersebut. Nilai f1-score adalah rata-rata harmonik dari presisi dan recall, yang menunjukkan keseimbangan antara keduanya. Nilai dukungan menunjukkan jumlah data yang sebenarnya memiliki kelas tersebut.

**b. Random Forest**

Random Forest adalah sebuah algoritma machine learning berdasarkan pada konsep ensemble learning, yaitu menggabungkan beberapa model yang lebih sederhana, yaitu pohon keputusan, untuk menghasilkan sebuah model yang lebih kuat dan akurat. Kelebihan dari algoritma ini adalah dapat mengurangi overfitting, meningkatkan stabilitas, dan menangani data yang memiliki fitur kategorikal dan numerik. Kekurangan dari algoritma ini adalah membutuhkan waktu dan memori yang lebih besar, sulit untuk diinterpretasikan, dan dapat menghasilkan hasil yang berbeda jika menggunakan nilai acak yang berbeda.

In [140]:
# Membuat objek rf yang merupakan sebuah estimator dari kelas RandomForestClassifier
rf = RandomForestClassifier()

# Melatih model rf dengan menggunakan data fitur dan target dari data latih
rf.fit(X_train, Y_train)

# Menggunakan model rf yang sudah dilatih untuk membuat prediksi kelas dari data fitur validasi, yaitu X_validation dan simpan hasilnya dalam variable predictions_rf
predictions_rf = rf.predict(X_validation)

# Mencetak string judul accuracy score dari random forest
print("Accuracy Score of Random Forest:")

# Mencetak nilai akurasi dari model rf, yaitu seberapa sering prediksi model rf sesuai dengan kelas yang sebenarnya dari data validasi
# Nilai akurasi dihitung dengan mengambil parameter Y_validation dan predictions_rf.
print(accuracy_score(Y_validation, predictions_rf))

# Mencetak string judul confusion matrix dari Random Forest
print("\nConfusion Matrix of Random Forest:")

# Mencetak confusion matrix dari model rf yang menampilkan jumlah data yang benar terklasifikasi (diagonal utama) dan jumlah data yang salah terklasifikasi (di luar diagonal utama) untuk setiap kelas yang ada
print(confusion_matrix(Y_validation, predictions_rf))

# Mencetak string judul classification report dari random forest
print("\nClassification Report of Random Forest:")

# Mencetak classification report dari model rf yang menampilkan nilai-nilai metrik evaluasi untuk setiap kelas, serta rata-rata dari semua kelas
print(classification_report(Y_validation, predictions_rf))

Accuracy Score of Random Forest:
0.8666666666666667

Confusion Matrix of Random Forest:
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]

Classification Report of Random Forest:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



**Penjelasan hasil:**

- Nilai akurasi adalah 0.8666666666666667, yang berarti model rf dapat memprediksi kelas yang benar untuk 87% dari data validasi

- Dari confusion matrix, dapat dilihat bahwa model rf dapat mengklasifikasikan semua data Iris-setosa dengan benar, tetapi ada beberapa kesalahan klasifikasi untuk data Iris-versicolor dan Iris-virginica. Misalnya, ada dua data Iris-versicolor yang diprediksi sebagai Iris-virginica, dan ada dua data Iris-virginica yang diprediksi sebagai Iris-versicolor.

- Dari classification report, dapat dilihat bahwa model rf memiliki nilai precision, recall, dan f1-score yang tinggi untuk kelas Iris-setosa. Nilai precision menunjukkan seberapa sering model rf memprediksi kelas yang benar dari semua prediksi yang dibuat untuk kelas tersebut. Nilai recall menunjukkan seberapa sering model rf memprediksi kelas yang benar dari semua data yang sebenarnya memiliki kelas tersebut. Nilai f1-score adalah rata-rata harmonik dari presisi dan recall, yang menunjukkan keseimbangan antara keduanya. Nilai dukungan menunjukkan jumlah data yang sebenarnya memiliki kelas tersebut.

**c. Gradient Boosting**

Gradient Boosting adalah sebuah algoritma machine learning berdasarkan pada konsep boosting, yaitu menggabungkan beberapa model yang lebih lemah, yaitu pohon keputusan, untuk menghasilkan sebuah model yang lebih kuat dan akurat. Kelebihan dari algoritma ini adalah dapat meningkatkan kinerja model secara bertahap, dapat menangani data yang memiliki fitur kategorikal dan numerik, dan dapat mengurangi bias. Kekurangan dari algoritma ini adalah rentan terhadap overfitting, membutuhkan waktu dan memori yang lebih besar, dan sulit untuk diinterpretasikan.

In [142]:
# Membuat objek gb yang merupakan sebuah estimator dari kelas GradientBoostingClassifier
gb = GradientBoostingClassifier()

# Melatih model gb dengan menggunakan data fitur dan target dari data latih
gb.fit(X_train, Y_train)

# Menggunakan model gb yang sudah dilatih untuk membuat prediksi kelas dari data fitur validasi, yaitu X_validation dan simpan hasilnya dalam variable predictions_gb
predictions_gb = gb.predict(X_validation)

# Mencetak string judul accuracy score dari gradient boosting
print("Accuracy Score of Gradient Boosting:")

# Mencetak nilai akurasi dari model gb, yaitu seberapa sering prediksi model gb sesuai dengan kelas yang sebenarnya dari data validasi
# Nilai akurasi dihitung dengan mengambil parameter Y_validation dan predictions_gb.
print(accuracy_score(Y_validation, predictions_gb))

# Mencetak string judul confusion matrix dari gradient boosting
print("\nConfusion Matrix of Gradient Boosting:")

# Mencetak confusion matrix dari model rf yang menampilkan jumlah data yang benar terklasifikasi (diagonal utama) dan jumlah data yang salah terklasifikasi (di luar diagonal utama) untuk setiap kelas yang ada
print(confusion_matrix(Y_validation, predictions_gb))

# Mencetak string judul classification report dari gradient boosting
print("\nClassification Report of Gradient Boosting:")

# Mencetak classification report dari model gb yang menampilkan nilai-nilai metrik evaluasi untuk setiap kelas, serta rata-rata dari semua kelas
print(classification_report(Y_validation, predictions_gb))

Accuracy Score of Gradient Boosting:
0.8666666666666667

Confusion Matrix of Gradient Boosting:
[[ 7  0  0]
 [ 0  9  3]
 [ 0  1 10]]

Classification Report of Gradient Boosting:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.90      0.75      0.82        12
 Iris-virginica       0.77      0.91      0.83        11

       accuracy                           0.87        30
      macro avg       0.89      0.89      0.88        30
   weighted avg       0.88      0.87      0.87        30



**Penjelasan hasil:**

- Nilai akurasi adalah 0.8666666666666667, yang berarti model gb dapat memprediksi kelas yang benar untuk 87% dari data validasi

- Dari confusion matrix, dapat dilihat bahwa model gb dapat mengklasifikasikan semua data Iris-setosa dengan benar, tetapi ada beberapa kesalahan klasifikasi untuk data Iris-versicolor dan Iris-virginica. Misalnya, ada tiga data Iris-versicolor yang diprediksi sebagai Iris-virginica, dan ada satu data Iris-virginica yang diprediksi sebagai Iris-versicolor.

- Dari classification report, dapat dilihat bahwa model gb memiliki nilai precision, recall, dan f1-score yang tinggi untuk kelas Iris-setosa. Nilai precision menunjukkan seberapa sering model gb memprediksi kelas yang benar dari semua prediksi yang dibuat untuk kelas tersebut. Nilai recall menunjukkan seberapa sering model gb memprediksi kelas yang benar dari semua data yang sebenarnya memiliki kelas tersebut. Nilai f1-score adalah rata-rata harmonik dari presisi dan recall, yang menunjukkan keseimbangan antara keduanya. Nilai dukungan menunjukkan jumlah data yang sebenarnya memiliki kelas tersebut.

**Kesimpulan:**

- Ketiga algoritma yang digunakan, yaitu Decision Tree, Random Forest, dan Gradient Boosting memiliki nilai akurasi yang sama yaitu 0.867, yang menunjukkan bahwa ketiga model klasifikasi dapat memprediksi kelas yang benar dari data validasi sebanyak 86.67% dari keseluruhan data. Nilai akurasi ini termasuk dalam kategori yang tinggi, sehingga dapat dikatakan bahwa ketiga model klasifikasi sesuai dengan data.

## **TUGAS NOMOR 2**

In [143]:
# Mengimpor modul Pandas dengan alias atau singkatan pd
import pandas as pd

## **1. Data yang digunakan**

In [144]:
# Mengimpor fungsi drive dari modul google.colab
from google.colab import drive

# Memanggil fungsi mount dari objek drive untuk mengaitkan sesi Colab dengan Google Drive
drive.mount('/content/drive')

# Menetapkan sebuah variabel bernama names yang berisi daftar nama-nama kolom yang terdapat pada dataset Rice
names2 = ['area', 'perimeter', 'major_axis_length', 'minor_axis_length', 'eccentricity', 'convex_area', 'extent', 'class']

# Membaca dataset dari file Excel sesuai dengan path dan datasetnya disimpan dalam variabel dataset2
dataset2 = pd.read_excel('/content/rice-dataset-commeo-and-osmancik/Rice_Dataset_Commeo_and_Osmancik/Rice_Cammeo_Osmancik.xlsx')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### **README:**
Untuk mengakses dataset, mohon terlebih dahulu mengupload file yang telah dilampirkan di github ke drive pribadi.

Kemudian ubah path dari file xlsxnya.

Terima kasih.

In [145]:
# Melihat 5 baris pertama dan 5 baris terakhir dari dataset
dataset2

Area   Perimeter  Major_Axis_Length  Minor_Axis_Length  Eccentricity  \
0     15231  525.578979         229.749878          85.093788      0.928882   
1     14656  494.311005         206.020065          91.730972      0.895405   
2     14634  501.122009         214.106781          87.768288      0.912118   
3     13176  458.342987         193.337387          87.448395      0.891861   
4     14688  507.166992         211.743378          89.312454      0.906691   
...     ...         ...                ...                ...           ...   
3805  11441  415.858002         170.486771          85.756592      0.864280   
3806  11625  421.390015         167.714798          89.462570      0.845850   
3807  12437  442.498993         183.572922          86.801979      0.881144   
3808   9882  392.296997         161.193985          78.210480      0.874406   
3809  11434  404.709991         161.079269          90.868195      0.825692   

      Convex_Area    Extent     Class  
0           15617  0.572896    Cammeo  
1           15072  0.615436    Cammeo  
2           14954  0.693259    Cammeo  
3           13368  0.640669    Cammeo  
4           15262  0.646024    Cammeo  
...           ...       ...       ...  
3805        11628  0.681012  Osmancik  
3806        11904  0.694279  Osmancik  
3807        12645  0.626739  Osmancik  
3808        10097  0.659064  Osmancik  
3809        11591  0.802949  Osmancik  

[3810 rows x 8 columns]

In [146]:
# Mengimpor modul model_selection dari library scikit-learn untuk membantu proses pemilihan dan evaluasi model machine learning
from sklearn import model_selection

**I. Membuat Valiadasi Dataset**

In [147]:
# Mengambil nilai-nilai dari DataFrame dataset1 dan dibuat menjadi sebuah array numpy dan disimpan pada sebuah variabel bernama array2
array2 = dataset2.values

# Assign semua baris dan tujuh kolom pertama pada variable X
X = array2[:,0:7]

# Assign semua baris dan kolom ke delapan pada variable Y
Y = array2[:,7]

# Menetapkan proporsi atau persentase data yang digunakan sebagai data validasi atau uji adalah 20% dan sisanya yaitu 80% sebagai data latih
validation_size = 0.20

# Menunjukkan nilai acak atau random state yang akan digunakan untuk memastikan hasil yang konsisten dan memastikan bahwa data akan dibagi dengan cara yang sama setiap kali kode dijalankan
seed = 7

# Menetapkan empat variable yang berisi hasil dari membagi data menjadi data latih dan  data validasi
# Fungsi model_selection.train_test_split memiliki parameter X dan Y yang merupakan data fitur dan target, text_size yang merupakan proporsi dan validasi, dan random_state yang merupakan nilai acak.
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size = validation_size, random_state=seed, shuffle=True)
# Kode di atas mengembalikan empat array yaitu :
# X_train dan Y_train yang merupakan data fitur dan target untuk data latih
# X_validation dan Y_validation yang merupakan data fitur dan target untuk data validasi

**II. Mencetak Ukuran atau Dimensi Data Latih dan Data Validasi**

In [148]:
# Mencetak string "X_train" dan ukuran dari array X_train yang berisi fitur untuk data latih
print("X_train :", X_train.shape)

# Mencetak string "X_validation" dan ukuran dari array X_validation yang berisi fitur untuk data validasi
print("X_validation :", X_validation.shape)

# Mencetak string "Y_train" dan ukuran dari array Y_train yang berisi data target untuk data latih
print("Y_train :", Y_train.shape)

# Mencetak string "Y_validation" dan ukuran dari array Y_validation yang berisi data target untuk data validasi
print("Y_validation :", Y_validation.shape)

X_train : (3048, 7)
X_validation : (762, 7)
Y_train : (3048,)
Y_validation : (762,)


**Penjelasan:**

- X_train : (3048, 7) menunjukkan bahwa data fitur untuk data latih memiliki 3048 baris dan 7 kolom

- X_validation : (762, 7) menunjukkan bahwa data fitur untuk data validasi memiliki 69 baris dan 6 kolom

- Y_train : (3048,) menunjukkan bahwa data  target untuk data latih memiliki 3048 baris dan 1 kolom

- Y_validation : (762,) menunjukkan bahwa data target untuk data validasi memiliki 762 baris dan 1 kolom

**III. K-Folds Cross Validation**

In [149]:
# Menunjukkan nilai acak atau random state yang digunakan untuk memastikan hasil yang konsisten dan memastikan bahwa data akan dibagi dengan cara yang sama setiap kali kode dijalankan
seed = 7

# Metrik evaluasi yang akan digunakan untuk membandingkan model yang berbeda adalah nilai akurasi, yaitu seberapa sering prediksi model sesuai dengan kelas yang sebenarnya dari data
scoring = 'accuracy'

**IV. Membangun Model**

In [150]:
# Mengimpor kelas KNeighborsClassifier dari modul sklearn.neigbors yaitu algoritma machine learning berdasarkan pada jarak anatara data dan tetangga terdekatnya
from sklearn.neighbors import KNeighborsClassifier

# Mengimpor kelas GaussianNB dari modul sklearn.naive_bayes yaitu algoritma machine learning yang berdasarkan pada teorema bayes dan mengasumsikan bahwa fitur memiliki distribusi normal atau gaussian
from sklearn.naive_bayes import GaussianNB

# Mengimpor kelas SVC dari modul sklearn.svm yaitu algoritma machine leraning berdasarkan konsep support vector machine yaitu mencari sebuah hyperplane yang dapat memisahkan data dengan margin terbesar
from sklearn.svm import SVC

In [151]:
# membuat sebuah variabel bernama models yang berisi sebuah daftar kosong
models = []

# Menambahkan sebuah elemen ke dalam daftar models yaitu tuple berisi string 'KNeighbors' dan sebuah objek dari kelas KNeighborsClassifier
models.append(('KNeighbors', KNeighborsClassifier()))

# Menambahkan sebuah elemen ke dalam daftar models yaitu tuple berisi string 'GaussianNB' dan sebuah objek dari kelas GaussianNB
models.append(('GaussianNB', GaussianNB()))

# Menambahkan sebuah elemen ke dalam daftar models yaitu tuple berisi string 'SVM' dan sebuah objek dari kelas SVC
models.append(('SVM', SVC()))

# Evaluate each model in turn

# Membuat variable result kosong
results = []

# Membuat varible names kosong
names = []

# Looping setiap elemen di models dan menetapkan variable name dan model sebagai nilai dari tuple yang ada di dalam elemen tsb
for name, model in models:
  # Membagi data menjadi 10 bagian atau folds dengan menggunakan nilai acak dari variable seed dan mengacak urutan data sebelum dibagi
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  # Membuat variable cv_result berisi sebuah array numpy yang berisi nilai akurasi dari model yang diuji dengan menggunakan metode cross-validation, dan parameter sbb:
  # - model sebagai model yang akan diuji
  # - X_train dan Y_train sebagai data fitur dan target dari data latih
  # - kfold sebagai metode pembagian data
  # - scoring sebagai metrik evaluasi yang digunakan
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)

  # Menambahkan nilai dari variable cv_result ke dalam daftar result
  results.append(cv_results)

  # Menambahkan nilai dari variable ke dalam daftar names
  names.append(name)

  # Membuat sebuah variable bernama msg yang ebrisi sebauh string berformat yang menampilkan nama model, nilai rata-rata , dan standar deviasi dari nilai akurasi yang ada di dalam variable cv_reslut
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())

  # Mencetak nilai dari variable msg ke layar
  print(msg)

KNeighbors: 0.886798 (0.014580)
GaussianNB: 0.909439 (0.013765)
SVM: 0.876629 (0.021134)


**Penjelasan:**

- KNN memiliki nilai akurasi 88.68% dan standar deviasi 1.46%
- NB memiliki nilai akurasi 90.94% dan standar deviasi 1.38%
- SVM memiliki nilai akurasi 87.66% dan standar deviasi 2.11%

**Kesimpulan:**
- NB memiliki nilai akurasi tertinggi dan dapat memprediksi kelas yang benar dari data validasi sebanyak 90.94% dari keseluruhan model data
- KNN memiliki nilai akurasi kedua tertinggi dan dapat memprediksi kelas yang benar dari data validasi sebanyak 88.68% dari keseluruhan data
- SVM memiliki nilai akurasi terendah dan dapat memprediksi kelas yang benar dari data validasi sebanyak 87.66% dari keseluruhan data
- Standar deviasi untuk ketiga algoritma kecil yang menunjukkan bahwa ketiga model tersebut relatif stabil dalam prediksi.

**V. Memilih Model Terbaik**

Berdasarkan hasil accuracy score diatas, NB (gaussian naive bayes) memiliki akurasi tertinggi.

In [152]:
# Mengimpor tiga fungsi dari modul sklearn.metrics yaitu accuracy_score, confusion_matrix, dan classification_report untuk menghitung dan menammpilkan beberapa metrik evaluasi untuk model machine learning
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [153]:
# Make predictions on validation dataset
# Membuat sebuah variable bernama nb yang berisi objek dari kelas GaussinNB
nb = GaussianNB()

# Melatih model nb dengan menggunakan data fitur dan target dari data latih yaitu X_train dan Y_train
nb.fit(X_train, Y_train)

# Membuat variable bernama predictions yang berisi sebuah array numpy yang berisi prediksi kelas dari model nb, untuk data fitur validasi yaitu Y_validation
predictions = nb.predict(X_validation)

# Mencetak string judul accuracy score dari nb
print("Accuracy Score of GaussianNB:")

# Mencetak nilai akurasi dari model nb yaitu seberapa sering prediksi model svm sesuai dengan kelas yang sebenarnya dari data validasi
print(accuracy_score(Y_validation, predictions))

# Mencetak string judul confusion matrix dari nb
print("\nConfusion Matrix of GaussianNB:")

# Mencetak confusion matrix dari model nb yang menampilkan jumlah data benar yang terklasifikasi (diagonal utama) dan jumlah data yang salah terklasifikasi (di luar diagonal utama) untuk setiap kelas yang ada
print(confusion_matrix(Y_validation, predictions))

# Mencetak string judul classification report dari nb
print("\nClassification Report of GaussianNB:")

# Mencetak classification report dari model nb yang menampilkan nilai-nilai precision, recall, f1-score, dan support untuk setiap kelas, serta rata-rata dari semua kelas
print(classification_report(Y_validation, predictions))

Accuracy Score of GaussianNB:
0.9199475065616798

Confusion Matrix of GaussianNB:
[[301  34]
 [ 27 400]]

Classification Report of GaussianNB:
              precision    recall  f1-score   support

      Cammeo       0.92      0.90      0.91       335
    Osmancik       0.92      0.94      0.93       427

    accuracy                           0.92       762
   macro avg       0.92      0.92      0.92       762
weighted avg       0.92      0.92      0.92       762



**Penjelasan hasil:**

- Nilai akurasi adalah 0.9199475065616798, yang berarti model GaussianNB dapat memprediksi kelas yang benar untuk 91.95% dari data validasi

- Dari confusion matrix, dapat dilihat bahwa model GaussianNB dapat mengklasifikasikan 301 data Cammeo dan 400 data Osmancik dengan benar. Terdapat beberapa kesalahan klasifikasi yaitu sebanyak 34 Cammeo diprediksi sebagai Osmancik dan 27 data Osmancik diprediksi sebagai Cammeo.

- Dari classification report, dapat dilihat bahwa model GaussianNB memiliki nilai precision, recall, dan f1-score yang termasuk tinggi untuk kedua kelas. Nilai precision menunjukkan seberapa sering model GaussianNB memprediksi kelas yang benar dari semua prediksi yang dibuat untuk kelas tersebut. Nilai recall menunjukkan seberapa sering model GaussianNB memprediksi kelas yang benar dari semua data yang sebenarnya memiliki kelas tersebut. Nilai f1-score adalah rata-rata harmonik dari presisi dan recall, yang menunjukkan keseimbangan antara keduanya. Nilai dukungan menunjukkan jumlah data yang sebenarnya memiliki kelas tersebut.